pkg/terra/client/client.go

```diff
@@ -0,0 +1,333 @@
+package client
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	"github.com/smartcontractkit/terra.go/tx"
+	rpchttp "github.com/tendermint/tendermint/rpc/client/http"
+
+	tmtypes "github.com/cosmos/cosmos-sdk/client/grpc/tmservice"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	txtypes "github.com/cosmos/cosmos-sdk/types/tx"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/smartcontractkit/terra.go/msg"
+	wasmtypes "github.com/terra-money/core/x/wasm/types"
+
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+
+	cosmosclient "github.com/cosmos/cosmos-sdk/client"
+	"github.com/pkg/errors"
+	"github.com/terra-money/core/app"
+
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+
+	"github.com/smartcontractkit/terra.go/key"
+)
+
+//go:generate mockery --name ReaderWriter --output ./mocks/
+type ReaderWriter interface {
+	Writer
+	Reader
+}
+
+// Only depends on the cosmos sdk types.
+type Reader interface {
+	GasPrice() sdk.DecCoin
+	Account(address sdk.AccAddress) (uint64, uint64, error)
+	ContractStore(contractAddress string, queryMsg []byte) ([]byte, error)
+	TxsEvents(events []string) (*txtypes.GetTxsEventResponse, error)
+	Tx(hash string) (*txtypes.GetTxResponse, error)
+	LatestBlock() (*tmtypes.GetLatestBlockResponse, error)
+	BlockByHeight(height int64) (*tmtypes.GetBlockByHeightResponse, error)
+	Balance(addr sdk.AccAddress, denom string) (*sdk.Coin, error)
+}
+
+type Writer interface {
+	// Assumes all msgs are for the same from address.
+	// We may want to support multiple from addresses + signers if a use case arises.
+	SignAndBroadcast(msgs []sdk.Msg, accountNum uint64, sequence uint64, gasPrice sdk.DecCoin, signer key.PrivKey, mode txtypes.BroadcastMode) (*txtypes.BroadcastTxResponse, error)
+	Broadcast(txBytes []byte, mode txtypes.BroadcastMode) (*txtypes.BroadcastTxResponse, error)
+	Simulate(txBytes []byte) (*txtypes.SimulateResponse, error)
+	SimulateUnsigned(msgs []sdk.Msg, sequence uint64) (*txtypes.SimulateResponse, error)
+	CreateAndSign(msgs []sdk.Msg, account uint64, sequence uint64, gasLimit uint64, gasPrice sdk.DecCoin, signer key.PrivKey, timeoutHeight uint64) ([]byte, error)
+}
```

Comment:
I'm curious: why the distinction between Reader and Writer here? 
A few things I can think of:
- you have a component that you want to make sure only reads or only writes. Is there one?
- you want to simplify mocking in tests. But afaict, you only generate the ReaderWriter mock and use that everywhere.
- you want to keep interfaces small and simple. They're not.

Reply:
yeah the whole terra relay only needs the reader actually. The terra txm in core is a reader/writer